### ISSUE TITLE
runtime weaver with java 8: java.lang.ClassCircularityError

### ISSUE THREAD

**Issue by nqzero** [kilim/kilim#46]
i'm trying to use the runtime weaver with java 8 (asm5-jdk8 branch). for some simple classes it works, but for others, such as kilim.examples.Tree, i get errors

```
ant clean compile
cp -R classes c1
ant weave
cp -R classes c2
rm -R c2/kilim/examples/
java -cp "c2:libs/*" -Dkilim.class.path=c1 kilim.tools.Kilim kilim.examples.Spawn
java -cp "c2:libs/*" -Dkilim.class.path=c1 kilim.tools.Kilim kilim.examples.Tree
```

Spawn works (prints "Hello from 0...)", Tree fails:

```
Exception in thread "main" java.lang.ClassCircularityError: kilim/examples/Tree
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2688)
    at java.lang.Class.getDeclaredMethods(Class.java:1962)
    at kilim.mirrors.RuntimeClassMirror.getDeclaredMethods(RuntimeClassMirrors.java:163)
    at kilim.mirrors.Detector.findPausableMethod(Detector.java:119)
    at kilim.mirrors.Detector.getPausableStatus(Detector.java:71)
    at kilim.analysis.MethodFlow.visitMethodInsn(MethodFlow.java:219)
    at org.objectweb.asm.MethodVisitor.visitMethodInsn(Unknown Source)
    at org.objectweb.asm.tree.MethodNode.visitMethodInsn(Unknown Source)
    at org.objectweb.asm.ClassReader.a(Unknown Source)
    at org.objectweb.asm.ClassReader.b(Unknown Source)
    at org.objectweb.asm.ClassReader.accept(Unknown Source)
    at org.objectweb.asm.ClassReader.accept(Unknown Source)
    at kilim.analysis.ClassFlow.analyze(ClassFlow.java:85)
    at kilim.analysis.ClassWeaver.weave(ClassWeaver.java:93)
    at kilim.tools.Weaver.weave(Weaver.java:297)
    at kilim.tools.Weaver.weave(Weaver.java:255)
    at kilim.WeavingClassLoader.findClass(WeavingClassLoader.java:61)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2688)
    at java.lang.Class.getDeclaredMethods(Class.java:1962)
    at kilim.mirrors.RuntimeClassMirror.getDeclaredMethods(RuntimeClassMirrors.java:163)
    at kilim.mirrors.Detector.findPausableMethod(Detector.java:119)
    at kilim.mirrors.Detector.getPausableStatus(Detector.java:71)
    at kilim.analysis.MethodFlow.visitMethodInsn(MethodFlow.java:219)
    at org.objectweb.asm.MethodVisitor.visitMethodInsn(Unknown Source)
    at org.objectweb.asm.tree.MethodNode.visitMethodInsn(Unknown Source)
    at org.objectweb.asm.ClassReader.a(Unknown Source)
    at org.objectweb.asm.ClassReader.b(Unknown Source)
    at org.objectweb.asm.ClassReader.accept(Unknown Source)
    at org.objectweb.asm.ClassReader.accept(Unknown Source)
    at kilim.analysis.ClassFlow.analyze(ClassFlow.java:85)
    at kilim.analysis.ClassWeaver.weave(ClassWeaver.java:93)
    at kilim.tools.Weaver.weave(Weaver.java:297)
    at kilim.tools.Weaver.weave(Weaver.java:255)
    at kilim.WeavingClassLoader.findClass(WeavingClassLoader.java:61)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    at kilim.tools.Kilim.main(Kilim.java:24)
```

**Comment by nqzero** [comment]
i've done a little bit of digging and it appears that WeavingClassLoader.findClass is calling the weaver, which in loads other classes (to determine if called methods are Pausable). before the first class finishes loading

so if in turn those classes depend on the first class, then everything deadlocks

@sriram-srinivasan does this sound right ? i'm thinking i might be able to work around the problem with a 2nd classloader. when we want to do reflection on the dependent classes, we first load them in the 2nd classloader and do the inspection there, then finish loading the original class in the WeavingClassLoader

**Comment by kilim** [comment]
Your reasoning sounds right, but I say this without any analysis at all. I’ll have to reload this stuff into my working memory before I venture another opinion. If you are able to find a fix, I’ll gladly take it!

**Comment by nqzero** [comment]
i tried the obvious solution yesterday - it handled the Tree.java example correctly, but failed on more complex hierarchies. spent most of the day reading thru the code and we're doing some non-intuitive stuff (the mirrors and delegates classes) that make changing classloaders non-trivial. i'll keep working on it - if i get to the point of giving up i'll let you know

**Comment by nqzero** [comment]
https://github.com/nqzero/kilim/tree/java8.rtweave2

runtime weaving (ie kilim.tools.Kilim) works for the examples and my port of jdk8 streams to kilim

i went thru the mirroring code and found a ton of stuff that wasn't actively being used, eg, the thread local Detector and the caching mirror, that made figuring out the code a lot more complicated. i deleted that stuff and then fixed the WeavingClassLoader by adding a proxy. if you take a look, all the classloader stuff is in the last commit (https://github.com/nqzero/kilim/commit/dee00430c19f26d9c47ea15c6d77e5fa9bec03ec)

not sure if i should backport the proxy to sriram's version or forward port the tools that i deleted to my version

first, i'm going to see if i can simplify the loader - right now it's classloader spaghetti. it would also be pretty easy to just weave from the classpath instead of working about -Dkilim.class.path (at the cost of slower startup - i haven't quantified this effect)

**Comment by nqzero** [comment]
https://github.com/nqzero/kilim/tree/srl.java8.formdata

specifically https://github.com/nqzero/kilim/commit/02f92eaad930b35ebb0997499fc9bf3776cf2f08 fixes this bug. i haven't figured out how to run the tests using rtweaving, but the examples run with kilim.tools.Kilim bootstrapping, as does my kilim-streams package

just cherry picking that commit should work, but there are other fixes on that branch that need to be integrated too

this patch just tweaks the existing code. however, i think the approach above might be better

* the thread-local Detector stuff is ugly and makes reasoning about the code much harder than it needs to be
* it's not clear that the caching class mirror is faster than just loading the classes in a proxy classloader
* the caching mirror results in class loading cycles with all sorts of corner cases ... it's impossible to test thoroughly
* the caching mirror results in a bunch of duplicate code

**Comment by kilim** [comment]
Thanks. I had inherited the detector stuff from the erjang project and have not put in a serious refactoring thought to the classloading issue. Classloaders are almost always fragile in Java.

**Comment by nqzero** [comment]
@sriram-srinivasan i sent an email to the kilim at malhar dot net address. if you don't have time to reply that's fine, but i'd like to know if you received it ? (the capchta acted as though the address might be abandoned)

**Comment by nqzero** [comment]
this is fixed in [kilim 2.0](https://github.com/nqzero/kilim) by a series of commits, the last of which is https://github.com/nqzero/kilim/commit/6c5e3bde3597a803de2c1cdb3e8b56488dcaf440

the log messages for most commits in this series begin with "rtweave"

this is the primary change in kilim 2.0 and makes for improved robustness with both runtime weaving and using AOT weaven code with alternative classloaders
